src/transformStateWithClones.js

```diff
@@ -7,7 +7,50 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const transformedStates = [];
+  let stateCopy = { ...state };
+
+  for (const action of actions) {
+    switch (action.type) {
+      case 'addProperties' :
+        stateCopy = addProperties(stateCopy, action);
+        transformedStates.push(stateCopy);
+        break;
+
+      case 'removeProperties' :
+        stateCopy = removeProperties(stateCopy, action);
+        transformedStates.push(stateCopy);
+        break;
+
+      case 'clear' :
+        stateCopy = {};
+        transformedStates.push(stateCopy);
+        break;
+
+      default :
+        return 'error';
+    }
+  }
+
+  return transformedStates;
+}
+
+function addProperties(state, action) {
+  const newState = { ...state };
+
+  Object.assign(newState, action.extraData);
+
+  return newState;
+}
+
+function removeProperties(state, action) {
+  const newState = { ...state };
```

Comment:
you don't need cloning states here
```suggestion
function addProperties(state, action) {
  Object.assign(state, action.extraData);

  return newState;
}

function removeProperties(state, action) {
```